### DISCUSSION TITLE
How to fetch only the client selected fields from the DB?

### DISCUSSION THREAD

**Discussion by Erythros** [MichalLytek/type-graphql#1316] [Q&A]
Let me start by saying that I'm not sure if this question should be linked to TypeGraphQL or to TypeORM. Maybe someone could help me better understand.

The problem (taken from the example directory):

```ts
@Entity()
@ObjectType()
export class Recipe {
  @Field(type => ID)
  @PrimaryGeneratedColumn()
  readonly id: number;

  @Field()
  @Column()
  title: string;

  @Field({ nullable: true })
  @Column({ nullable: true })
  description?: string;

  @Field(type => [Rate])
  @OneToMany(type => Rate, rate => rate.recipe, { cascade: ["insert"] })
  ratings: Rate[];

  @Field(type => User)
  @ManyToOne(type => User)
  author: User;
  @RelationId((recipe: Recipe) => recipe.author)
  authorId: number;
}
```

If I want to fetch a recipe by ID (including the author and the ratings), I'd have to do something like this
```ts
    const recipe = await this.recipeRepository.findOne(rateInput.recipeId, {
      relations: ["ratings", "author"],
    });
```

That's great if the client wants everything. However, if the client **only** needs the `author`, but not the `ratings`, my Query would still fetch the ratings from the DB, it would just not return it. 
Maybe the client only wants the author ID, not the entire user object, etc.

Alternatively, I can use the `QueryBuilder` and do joins on the user/rating. But the same issue occurs since the joins are not done dynamically based on what the client wants/selects.

Is there a way to only fetch the fields from the DB that the client wants?
Otherwise, the main GraphQL advantage of "solving under or overfetching" is kinda wasted for me...

Note: When using TypeGraphQL with Prisma, [there is a plugin](https://paljs.com/plugins/select/) that helps solve this issue. However, I couldn't find any solution for TypeGraphQL + TypeORM.

**Comment by hkd987** [comment]
You will need to use the info method and dynamically build your query based on the data the client asks for. We currently do this in production, but I can't share much more than that.

**Reply by kasir-barati** [reply]
No need to go through that much manual labor, just use [join-monster](https://github.com/join-monster/join-monster), it will generate a SQL query based on the query.

**Comment by karladler** [comment]
Have look at [type-graphql-data](https://github.com/slaypni/type-graphql-dataloader) loader which is easy to set up and also solves the [GraphQL N+1 problem](https://medium.com/the-marcy-lab-school/what-is-the-n-1-problem-in-graphql-dd4921cb3c1a).

Once set up, in most cases, you just need to add `@TypeormLoader()` decorator to your relations.

**Reply by Erythros** [reply]
@karladler that's a cool idea. The problem is that this library is already out of date and does not support TypeORM 0.3+.
It's not something one would like to use in production.

But yea, I'll do a  deeper dive into it, see what I can learn. 

Would be nice if TGQL would provide this built-in one day...

**Reply by MichalLytek** [reply]
👀 #10

**Reply by karladler** [reply]
> 👀 #10

I think the issues is related to what the question from the headline suggests, but the actual question is about querying relations which seems to be a perfect use case for data loader. I'm using it currently with typeorm < 0.3 and I hope there will be a solution for using it with typeorm > 0.3 soon.

**Reply by karladler** [reply]
@Erythros it seems to work with a small tweak: https://github.com/slaypni/type-graphql-dataloader/issues/34
